html5/js/OffscreenDecodeWorker.js

```diff
@@ -418,81 +158,68 @@ class WindowDecoder {
           sw,
           sh
         );
-        if (this.debug) {
-          this.paint_box(context, "brown", sx + xdelta, sy + ydelta, sw, sh);
-        }
+        this.paint_box(coding, context, sx, sy, sw, sh);
       }
-    } else if (coding == "frame") {
+    } else if (coding.startsWith("frame")) {
       let enc_width = width;
       let enc_height = height;
       const options = packet[10] || {};
       const scaled_size = options["scaled_size"];
-      if (scaled_size) {
+      if (scaled_size && (enc_width > width || enc_height > height)) {
         enc_width = scaled_size[0];
         enc_height = scaled_size[1];
       }
       context.drawImage(data, x, y, enc_width, enc_height);
       data.close();
-      paint_box();
-    } else if (coding == "throttle") {
-      //we are skipping this frame
-    } else if (coding == "void") {
-      //nothing to do
-    } else {
-      this.decode_error(packet, `unsupported encoding: ${coding}`);
+      this.paint_box(coding, context, x, y, enc_width, enc_height);
     }
+
+    // Decode ok.
     const options = packet[10] || {};
-    const flush = options["flush"] || 0;
-    if (flush == 0 && context.commit) {
-      context.commit();
-    }
+    const decode_time = Math.round(1000 * (performance.now() - start));
+    options["decode_time"] = Math.max(0, decode_time);
+    packet[6] = "offscreen-painted";
+    packet[7] = null;
+    packet[10] = options;
+    self.postMessage({ draw: packet, start });
   }
 
-  paint_box(context, color, px, py, pw, ph) {
-    if (color) {
-      context.strokeStyle = color;
+  paint_box(coding, context, px, py, pw, ph) {
+    if (!this.debug) {
+      return;
+    }
+    const source_encoding = coding.split(":")[1] || ""; //ie: "rgb24"
```

Comment:
I like it!
Much easier / better than keeping track of the original encoding in the long paint function.